C++: Query for multiplications used in allocations.

geoffw0 opened the pull request:
Experimental and rather quickly put together query for (non-constant) multiplication results that are used as the size for an allocation.  This is a practically motivated query intended to be a fast way to find a common pattern for potential vulnerabilities - e.g. it showed some promise on VLC.

Related:
 - cpp/integer-multiplication-cast-to-long "Multiplication result converted to larger type"
 - cpp/uncontrolled-allocation-size "Overflow in uncontrolled allocation size"

TODO now:
- [x] review real world results (for potential; they don't need to be perfect for experimental) and set query precision
- [x] write a proper violation message and query headers
- [x] add some tests
- [x] add basic qhelp with an example

TODO later (i.e. not required for merging into experimental):
- [ ] apply range analysis on the multiplication to build confidence it may actually overflow? (@mathiasvp said: "There’s nothing inherently wrong with the result of a multiplication flowing into an allocation, so it would probably need to be more constrained." ... "If we preserve interesting results when we restrict the multiplications to be potentially overflowing I think it could be a good experimental query")
- [ ] apply taint tracking on the multiplication to ensure its user controlled??? (but this might just turn the query results into a subset of cpp/uncontrolled-allocation-size, i.e. remove many of the interesting results)
- [ ] we've also looked at similar cases that flow into a call to `memset`.  In the long run, perhaps 'allocation' could be generalized to something like 'size of any buffer operation'?

@jbj I'm interested in your thoughts on this.  Note that the target is only experimental, at least for now.

MathiasVP commented on cpp/ql/src/experimental/Security/CWE/CWE-190/AllocMultiplicationOverflow.ql:
```diff
@@ -0,0 +1,42 @@
+    // a multiplication of two non-constant expressions
+    exists(MulExpr me |
+      me = node.asExpr() and
+      forall(Expr e | e = me.getAnOperand() | not exists(e.getValue()))
```
This should be equivalent, and AFAIK could end up being optimized better.
```suggestion
      not exists(Expr e | e = me.getAnOperand() and exists(e.getValue()))
```

geoffw0 replied:
What is the intuition for it being better optimized?

MathiasVP replied:
Hm. Looking at my notes I see now that I was wrong - I think it doesn't make any difference when the body of the `forall` doesn't involve recursion, so feel free to choose whatever form you find most readable.

geoffw0 replied:
Out of interest, what is the issue when the body involves recursion?  (assuming negative recursion doesn't crop up)

MathiasVP replied:
It's an example of non-linear recursion, which is slow because the recursive call cannot just be compiled down to a join with the `prev_delta` of the previous iteration. It's briefly mentioned in https://docs.google.com/document/d/12LURUTA51n1xeuoyGHhnSNTTjIa-1LQFG2sb_i-WhNI/edit#.

geoffw0 commented:
Looking at [a small sample of results on LGTM](https://lgtm.com/query/5389244427323910384/), I think there are quite a lot of false positives where the numbers come from things that are probably known to be small.  I've set the `@precision` of the query to `low` for now - and I think we'll need to add range analysis or perhaps taint tracking or something similar before this query could possibly come out of experimental.

geoffw0 commented:
Added qhelp and changed this from a draft PR to 'ready for review'.

I plan to create a Jira issue for the points about improving results, and potentially moving this query out of experimental, but what we have now is useful (as proved in the hackathon) and complete.

geoffw0 commented:
@hubwriter please could I have a quick documentation review.  For context, this is only being merged into 'experimental' (which has lower standards than other queries) for now, but it may be moved out in future so I'd prefer the qhelp be up to scratch.

hubwriter reviewed (approved):
query help LGTM 👍

geoffw0 commented:
I believe this PR can be merged.

MathiasVP reviewed (approved):
Indeed, this query (as is) has demonstrated its value in the hackathon. Let's investigate the changes needed to bring it out of experimental in another PR.

geoffw0 commented:
Created an issue for further improvements: https://github.com/github/codeql-c-analysis-team/issues/203

intrigus-lgtm commented:
> Indeed, this query (as is) has demonstrated its value in the hackathon.

I assume this was an internal hackathon?

MathiasVP commented:
> I assume this was an internal hackathon?

Yep. (Sorry for not making that clear in my comment.)
